src/lib/support/tests/TestBytesCircularBuffer.cpp

```diff
@@ -0,0 +1,191 @@
+    NL_TEST_ASSERT(inSuite, buffer.ReadFront(readSpan) == CHIP_NO_ERROR);
+    NL_TEST_ASSERT(inSuite, memcmp(s2, readSpan.data(), 4) == 0);
+
+    // Used = 9, Free = 0
+    NL_TEST_ASSERT(inSuite, buffer.Push(ByteSpan(s1, 1)) == CHIP_NO_ERROR);
+    NL_TEST_ASSERT(inSuite, !buffer.IsEmpty());
+    NL_TEST_ASSERT(inSuite, buffer.GetFrontSize() == 4);
+    NL_TEST_ASSERT(inSuite, buffer.ReadFront(readSpan) == CHIP_NO_ERROR);
+    NL_TEST_ASSERT(inSuite, memcmp(s2, readSpan.data(), 4) == 0);
+
+    // Used = 3, Free = 6
+    NL_TEST_ASSERT(inSuite, buffer.Pop() == CHIP_NO_ERROR);
+    NL_TEST_ASSERT(inSuite, !buffer.IsEmpty());
+    NL_TEST_ASSERT(inSuite, buffer.GetFrontSize() == 1);
+    NL_TEST_ASSERT(inSuite, buffer.ReadFront(readSpan) == CHIP_NO_ERROR);
+    NL_TEST_ASSERT(inSuite, memcmp(s1, readSpan.data(), 1) == 0);
+}
+
+void TestPushWrapAtEdge(nlTestSuite * inSuite, void * inContext)
+{
+    uint8_t storage[10];
+    const uint8_t s1[] = { '1', '2', '3', '4', '5', '6', '7', '8', '9', '0' };
+    const uint8_t s2[] = { 'a', 'b', 'c', 'd', 'e', 'f', 'g', 'h', 'i', 'j' };
+    uint8_t o[sizeof(s1)];
+    MutableByteSpan readSpan(o);
+
+    // Total space = 9
+    BytesCircularBuffer buffer(storage, sizeof(storage));
+    NL_TEST_ASSERT(inSuite, buffer.IsEmpty());
+
+    // Used = 6, Free = 3
+    NL_TEST_ASSERT(inSuite, buffer.Push(ByteSpan(s1, 4)) == CHIP_NO_ERROR);
+    NL_TEST_ASSERT(inSuite, !buffer.IsEmpty());
+    NL_TEST_ASSERT(inSuite, buffer.GetFrontSize() == 4);
+    NL_TEST_ASSERT(inSuite, buffer.ReadFront(readSpan) == CHIP_NO_ERROR);
+    NL_TEST_ASSERT(inSuite, memcmp(s1, readSpan.data(), 4) == 0);
+
+    // Used = 9, Free = 0, mDataEnd wrap to 0
+    NL_TEST_ASSERT(inSuite, buffer.Push(ByteSpan(s2, 1)) == CHIP_NO_ERROR);
+    NL_TEST_ASSERT(inSuite, !buffer.IsEmpty());
+    NL_TEST_ASSERT(inSuite, buffer.GetFrontSize() == 4);
+    NL_TEST_ASSERT(inSuite, buffer.ReadFront(readSpan) == CHIP_NO_ERROR);
+    NL_TEST_ASSERT(inSuite, memcmp(s1, readSpan.data(), 4) == 0);
+
+    // Used = 3, Free = 6
+    NL_TEST_ASSERT(inSuite, buffer.Pop() == CHIP_NO_ERROR);
+    NL_TEST_ASSERT(inSuite, !buffer.IsEmpty());
+    NL_TEST_ASSERT(inSuite, buffer.GetFrontSize() == 1);
+    NL_TEST_ASSERT(inSuite, buffer.ReadFront(readSpan) == CHIP_NO_ERROR);
+    NL_TEST_ASSERT(inSuite, memcmp(s2, readSpan.data(), 1) == 0);
+
+    // All freed
+    NL_TEST_ASSERT(inSuite, buffer.Pop() == CHIP_NO_ERROR);
+    NL_TEST_ASSERT(inSuite, buffer.IsEmpty());
+
+    // Try normal push/pop again
+    TestPushWrap(inSuite, inContext);
+}
+
+int Setup(void * inContext)
+{
+    return SUCCESS;
+}
+
+int Teardown(void * inContext)
+{
+    return SUCCESS;
+}
+
+} // namespace
+
+#define NL_TEST_DEF_FN(fn) NL_TEST_DEF("Test " #fn, fn)
+/**
+ *   Test Suite. It lists all the test functions.
+ */
+static const nlTest sTests[] = { NL_TEST_DEF_FN(TestPushPop), NL_TEST_DEF_FN(TestPushInvalid), NL_TEST_DEF_FN(TestPushWrap),
```

Comment:
To repeat my previous comment, adjusted for the parts that got addresse, please add tests for at least the following:

1. What happens when trying to push something that would, when added to what's there already, exactly equal mCapacity in total size.  I don't see any tests for this here.
2. Explicitly exercise the case when mDataEnd would end up equaling mCapacity (and should wrap to 0).
3. Exercise the case when mDataStart ends up equal to mCapacity (and should wrap to 0). That is, a case when mDataEnd became 0, then we pushed something, then we popped all the things that used to be there.

Plus the test that would catch the `Push` issue above....